File: src/Engine/Pactor.java
package Engine;

import functionpointers.VoidFunctionPointer;
import InternalInterfaces.AttributeSetter;
import InternalInterfaces.PactorCollisionFunction;
import PacDaddyApplicationInterfaces.PacDaddyAttributeReader;

final public class Pactor extends Actor implements PacDaddyAttributeReader, AttributeSetter {

	final private GameAttributes attributes;
	private PactorCollisionFunction onCollision;
	
	public Pactor() {
		attributes = new GameAttributes();
		onCollision = PactorCollisionFunction.EMPTY_FUNCTION;
		learnBasics();
	}
	
	public void setOnCollisionFunction(PactorCollisionFunction ON_COLLISION) {
		onCollision = ON_COLLISION;
	}
	
	public void setAttribute(String name, Object value) {
		attributes.setAttribute(name, value);
	}
	
	public Object getValueOf(String attributeName) {
		return attributes.getValueOf(attributeName);
	}

	public String[] getAttributes() {
		return attributes.getAttributes();
	}
	
	void notifyCollidedWith(Pactor collidedWith) {
		onCollision.call(collidedWith);
	}
	
	private void requestDirection(String direction) {
		if (getValueOf("DIRECTION") != direction) {
			setAttribute("REQUESTED_DIRECTION", direction);
		}
	}
	
	private void learnBasics() {
		setAttribute("DIRECTION", "NONE");
		setAttribute("REQUESTED_DIRECTION", "NONE");
		learnAction("UP", new VoidFunctionPointer() {
			public void call() { requestDirection("UP"); }
		});
		learnAction("DOWN", new VoidFunctionPointer() {
			public void call() { requestDirection("DOWN"); }
		});
		learnAction("LEFT", new VoidFunctionPointer() {
			public void call() { requestDirection("LEFT"); }
		});
		learnAction("RIGHT", new VoidFunctionPointer() {
			public void call() { requestDirection("RIGHT"); }
		});
		learnAction("NONE", new VoidFunctionPointer() {
			public void call() { }
		});
	}
	
}

File: PacDaddyGameWrapper/Pactor.lua
--require("luasrc/VoidFunctionPointer")
local public = {}

--local function requestDirection(pactor, direction)
--    if pactor:getValueOf("DIRECTION") ~= direction then
--        pactor:setAttribute("REQUESTED_DIRECTION", direction)
--    end
--end
--
--local function learnBasics(pactor)
--    pactor:setAttribute("DIRECTION", "NONE")
--    pactor:setAttribute("REQUESTED_DIRECTION", "NONE")
--    pactor:learnAction("UP",    VoidFunctionPointer(function() requestDirection(pactor, "UP") end))
--    pactor:learnAction("DOWN",  VoidFunctionPointer(function() requestDirection(pactor, "DOWN") end))
--    pactor:learnAction("LEFT",  VoidFunctionPointer(function() requestDirection(pactor, "LEFT") end))
--    pactor:learnAction("RIGHT", VoidFunctionPointer(function() requestDirection(pactor, "RIGHT") end))
--    pactor:learnAction("NONE",  VoidFunctionPointer(function() end))
--end

local function new()
    local pactor = luajava.newInstance("Engine.Pactor")
--    learnBasics(pactor)
    return pactor
end

public.new = new

return public

